pkg/operator/controller/externaldns/deployment.go

```diff
@@ -0,0 +1,274 @@
+		return true, current, err
+	}
+
+	// getting the deployment from API in case a third party updated it after the update
+	return r.currentExternalDNSDeployment(ctx, namespace, externalDNS)
+}
+
+// currentExternalDNSDeployment gets the current externalDNS deployment resource.
+func (r *reconciler) currentExternalDNSDeployment(ctx context.Context, namespace string, externalDNS *operatorv1alpha1.ExternalDNS) (bool, *appsv1.Deployment, error) {
+	depl := &appsv1.Deployment{}
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: namespace, Name: externalDNS.Name}, depl); err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil, nil
+		}
+		return false, nil, err
+	}
+	return true, depl, nil
+}
+
+// desiredExternalDNSDeployment returns the desired deployment resource.
+func desiredExternalDNSDeployment(namespace, image string, serviceAccount *corev1.ServiceAccount, externalDNS *operatorv1alpha1.ExternalDNS) (*appsv1.Deployment, error) {
+	replicas := int32(1)
+
+	matchLbl := map[string]string{
+		appNameLabel:     controller.ExternalDNSBaseName,
+		appInstanceLabel: externalDNS.Name,
+	}
+
+	nodeSelectorLbl := map[string]string{
+		osLabel:             linuxOS,
+		masterNodeRoleLabel: "",
+	}
+
+	depl := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      controller.ExternalDNSResourceName(externalDNS),
+			Namespace: namespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: matchLbl,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: matchLbl,
+				},
+				Spec: corev1.PodSpec{
+					ServiceAccountName: serviceAccount.Name,
+					NodeSelector:       nodeSelectorLbl,
+				},
+			},
+		},
+	}
+
+	for i, zone := range externalDNS.Spec.Zones {
+		container, err := populateExternalDNSContainer(image, metricsStartPort+i, zone, externalDNS)
+		if err != nil {
+			return nil, err
+		}
+		depl.Spec.Template.Spec.Containers = append(depl.Spec.Template.Spec.Containers, *container)
+	}
+
+	return depl, nil
+}
+
+// WIP function for generating container specs for one container at a time.
+func populateExternalDNSContainer(image string, metricsPort int, zone string, externalDNS *operatorv1alpha1.ExternalDNS) (*corev1.Container, error) {
+	name := "externaldns-" + zone
+
+	args := []string{
+		fmt.Sprintf("--metrics-address=127.0.0.1:%d", metricsPort),
+		fmt.Sprintf("--txt-owner-id=externaldns-%s", externalDNS.Name),
+		fmt.Sprintf("--zone-id-filter=%s", zone),
+		"--policy=sync",
+		"--registry=txt",
+		"--log-level=debug",
+	}
+
+	provider, ok := providerStringTable[externalDNS.Spec.Provider.Type]
+	if !ok {
+		return nil, fmt.Errorf("%s is not a supported provider", externalDNS.Spec.Provider.Type)
+	}
+	args = append(args, fmt.Sprintf("--provider=%s", provider))
+
+	//TODO: Add provider credentials logic
+
+	source, ok := sourceStringTable[externalDNS.Spec.Source.Type]
```

Comment:
The caller could do the source lookup, same way and with the same benefits that I described in my comment about the provider lookup.

Reply:
You mean you'd prefer to see something like:
```golang
func buildExternalDNSContainer(image, zone, provider string, source *ExternalDNSSource)
```
as a declaration instead of passing the whole `ExternalDNS` instance?

Reply:
I'm just suggesting that the `sourceStringTable[externalDNS.Spec.Source.Type]` lookup could be in `desiredExternalDNSDeployment`, and then `buildExternalDNSContainer` couldn't get an error.  If `desiredExternalDNSDeployment` did both the provider and source map lookups, then `buildExternalDNSContainer` would never return a non-nil error value, and the error return value could be deleted.  (Maybe as more options are added, more reasons to return an error will arise though?)

Reply:
Ah ok, I see. As a matter of fact the provider mapping can be replaced with a simple `ToLower()` as the values are already validated by OpenAPI of CRD and their ExternalDNS's equivalents are the same but in lower case.
The source unfortunately has 1 option which is drastically different in API and ExternalDNS: `OpenShiftRoute` vs `openshift-route`. The rest is also just a different case.
Well, ok, I'll pass them both as params, this will make the function a pure builder with no error handling, we'll see if this will work out (number of input params vs error output param).